Starting to add more general python.

mrshirts opened the pull request:
A first pass at more general python. Will need tweaking to be intelligible.

The original MSM case was pretty good (with some MRS: comments in below). 

What I am trying to do is push the gmxapi dependence down a layer (underneath a gromacs_wrapper), so we could swap out other programs (todo: write a lammps_wrapper that implements the same capabilities as gmxapi as called through command line tools).  The wrappers should have essentially the same methods so that upper layers can interact with any MD simulation engine transparently, with memory costs and speed beinig the prime tradeoff.

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# that contains all information needed to star the simulation.
+# 
+# So if one has N simulations, you would start with a list of N input objects.
+# Each object could be tagged as a package of a supported type (GROMACS, LAMMPS, NAMD),
```
In this case, `make_input` is actually, `gromacs_wrapper.make_input`, which indicates to the user that it accepts the GROMACS-specific set of input files, and allows for a module-specific object to be produced that we can either convert for other tools or optimize for GROMACS tool consumers. We could add a layer on top that could use a single command and more uniform or generalized set of key words, plus a key word indicating the simulation platform, which would dispatch to tools for known simulators. But I think this would actually reduce utility, since it would fail at run time instead of while checking syntax, and it is at least as clear to a user to import `make_input` from `gromacs` as it is to specify `gromacs` as an argument or function name.

mrshirts replied:
OK, that makes sense: each make_input is a function of a XXXX_wrapper module, so it clearly takes inputs of program XXXXX, and is run by the XXX_wrapper.simulate program of that.

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# and wrapper code for each supported type knows how to unwrap it.
+#
+# for GROMACS, it would be a .tpr file (or internal represntation of
+# the .tpr file) For LAMMPS, it's the array of input files, since
```
I think it would be either a gmxapi.SimulationInput object or a lammps.SimulationInput object, each of which can support some amount of a generic scalems.SimulationInput interface.

mrshirts replied:
OK, we will need to talk more about this.

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# MRS: I'm not sold on using the name subgraph, because it doesn't
+#      describe what it is a subgraph of. Can this be made more specific?
+
+# MRS: can a name be used instead of 'P' below? 
```
How about `transition_matrix`?

mrshirts replied:
Yes, better!

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# and can be used in a control operation.
+
+# MRS: I'm not sold on using the name subgraph, because it doesn't
+#      describe what it is a subgraph of. Can this be made more specific?
```
We document that work flows are represented by directed acyclic graphs, so `subgraph` is the most literal description of what the command is creating. A more generic term could be "fused operation." The symbol to which the subgraph is assigned could more informatively be named `simulation_and_analysis_iteration` or something.

mrshirts replied:
> A more generic term could be "fused operation."

I was sort of thinking of a LESS generic term, to be more specific. But naming the output symbol help interpretability a lot.

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+    # In the default work graph, add a node that depends on `condition` and
+    # wraps subgraph.
+
+    my_loop = gmx.while_loop(operation=subgraph,
```
Lines 84-89 should be unindented.

mrshirts replied:
Done.

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+        'P': scalems.ndarray(0., shape=(N, N)),
+        'is_converged': False})
+
+with subgraph:
```
I have come to prefer a syntax based on defining a (sub) class instead of hiding a subgraph builder as a context manager. Did I share an example of such an alternative? It might be clearer, too, in the sense that the user would be defining a new operation/command/function in the form of a subgraph/fusion of other standard command syntax. Thoughts?

mrshirts replied:
Maybe we can talk about this?

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+import gmxapi
+import scalems
+
+# MRS: we probably want to make these parameters keywords parameters
```
Keyword works for me, I like the more explicit definitions.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+    def __init__(self, topfile, trajectory, P, N):
+
+        # Build markov model with PyEmma
+        feat = coor.featurizer(topfile)  # MRS: should be using something more generalizable than this.
```
I looked into pyemma more, and saw that topfile refers to a molecular topology file which can just be a pdb file or a mdtraj.Topology instance; not a gromacs topology file.   Changed to molecular_topology_file to be less ambiguous.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# name, or list of file names if multiple inputs are needed to
+# intitialize the simulation.
+#
+# In theory, this could be abstracted further as some sort of input bundle.
```
OK, I need to think more about broadcasting to shapes based on the shape of the input data.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# Assuming MSMAnalyzer is an existing tool we do not want to modify,
+# create a scalems compatible operation by wrapping with a provided utility.
+
+#MRS: suggest names rather than 'P' and 'N'
```
OK , some potential confusion between which N is number of clusters and number of simulations, I will think about it.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+initial_input = scalems.broadcast(initial_simulation_input, shape=(N,))
+
+# We will need a pdb for MSM building in PyEmma
+initial_pdb = coordinate_inputs[0]
```
This was left over from the initial draft.  I think the initial is just to be able to build the molecular topology, but I'll have to think more about it.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# MRS: seems like some instructions would need to be given here, such as:
+#       # which entries in the files to replaced by random numbers
+#       # may need a temporary directory to store modified files.
+#       # so maybe the broadcasating would be done in make_input
```
I'll need to think a bit about broadcasting here.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+# Set up an array of N simulations, starting from a single input.
+# MRS: seems like some instructions would need to be given here, such as:
+#       # which entries in the files to replaced by random numbers
+#       # may need a temporary directory to store modified files.
```
Agreed, though they should be able to eventually find the files if needed for debugging.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+
+# Set up an array of N simulations, starting from a single input.
+# MRS: seems like some instructions would need to be given here, such as:
+#       # which entries in the files to replaced by random numbers
```
Right, agreed that is where what entries are specified are changed.

mrshirts commented on docs/applications_general.py:
```diff
@@ -0,0 +1,211 @@
+    # terminate the simulation
+    allframes = collect_configurations(md.output.trajectory)
+    
+    adaptive_msm = analysis.msm_analyzer(topfile=editconf.file['-o'],
```
Or does it need to be done at all?  Not clear yet to me.

eirrgang replied:
Re-posing as a slightly different question: Where does the code lives that converts the `.gro` or `.tpr` file provided by a user into the `.pdb` file required by PyEmma?

mrshirts replied:
Probably should belong to the wrapper tools associated with the simulation program - convert_internal_structure_to_pdb

eirrgang commented on docs/applications_general.py:
```diff
@@ -37,10 +27,10 @@
 
 # Set up an array of N simulations, starting from a single input.
```
To clarify: In the initial version of the script I posted, this comment was intended to indicate that these two commands converted a single simulation input to N identical simulation inputs. Substitutions could be performed by modifying the N data streams, or by using different commands (not shown) that produce N non-identical data streams.

eirrgang commented on docs/applications_general.py:
```diff
@@ -52,27 +42,24 @@
+simulation_and_analysis_iteration = scalems.subgraph(variables={
         'conformation': initial_input,
-        'P': scalems.ndarray(0., shape=(N, N)),
+        'transition_matrix': scalems.ndarray(0., shape=(N, N)),  # number of simulations or number of clusters?
```
Oops. Yes, here N means number of simulations, but seems to be used to set the number of cluster, too. I think the idea was that it would be convenient (after the first iteration) to be running one simulation per cluster at each iteration and then re-clustering the conformations collected from all simulations, then resampling from the new clusters for the next round of input.

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,190 @@
+    # terminate the simulation
+    allframes = collect_configurations(md.output.trajectory)
+    
+    adaptive_msm = analysis.msm_analyzer(molecular_topology=initial_pdb,
```
To do: consider how to express this more generically in terms of accumulating some number of samples from some number of simulations to cluster some number of model states to propose some number of follow-up simulations, where these numbers are decoupled and where the model accumulates data. Consider the constraints, if any, on whether batches of samples can be provided in arbitrary order or in arbitrary sized batches. This then leads to a generalization such that the algorithm can be expressed in various asynchronous ways, and we can explore iterator patterns.

mrshirts commented:
OK, cleaned it up a bit.  Thoughts so far?  Did you want to give a version that iterated differently?

mrshirts commented:
What is the next thing to do?
* Fill out more of the operations of the wrapper objects?  
* Start doing more things with MSM builder?
* You mentioned adding more asynchronous capability?

eirrgang commented on docs/applications_general.py:
```diff
@@ -9,45 +9,39 @@
+# NOTE: we presume that broadcast creates num_simulations even if the initial simulation input contains
+# fewer initial coordinates than num simulations. Like, if num_simulations is 40, and the initial 
+# coordinate list contains 7, then it will iterate (1,2,3,4,5,6,7,1,2,3,4,5, ... until reaching 40)?
+# and so on with the other files?
```
That would not be consistent with typical meanings of "broadcast." In particular, I think we should have behavior as close to numpy as possible, though we may need to make a few decisions about implicit behavior or no-op sorts of things, but in general, I think that if you have an object of shape `(n,)` and you broadcast to dimensionality `(m,)` you would get an object of shape `(m,n)`, but we may want to make it an error unless the target dimensionality is either explicitly `(m,n)` or `(m,PLACEHOLDER)` where `PLACEHOLDER` is something we invent to stand in for a dimension size we don't know a priori. I think it is okay to let `(n,)` broadcast to `(m,n)` with a directive of an `m`-dimensional broadcast, but we can adopt a different syntax for the broadcast shape if that is confusing.

mrshirts replied:
OK, that makes sense I was just trying to understand the intended behavior in different situations.

eirrgang commented on docs/applications_general.py:
```diff
@@ -166,18 +159,19 @@ def make_input(simulation_parameters = ['md.mdp'],
     # modify that input object, and convert the simulataon to back into a simulation inpput object.
     # the pattern holds. 
-    # key thing is that the overall program flow.  You package the iniputs and pass iit to th next tools. 
+    # key thing is that the overall program flow.  You package the iniputs and pass it to the next tools. 
```
sp

eirrgang commented on docs/applications_general.py:
```diff
@@ -166,18 +159,19 @@ def make_input(simulation_parameters = ['md.mdp'],
+    #
     # add features that if it is expecting 1 tpr and you give it 10 tpr files.
-    # it creates a simiulatioin iniput object that. What is specified.  If you use it with other gmxapi components.
+    # it creates a simulatioin input object that has dynamic shape If you use it with other gmxapi components.
```
sp

eirrgang commented on docs/applications_general.py:
```diff
@@ -0,0 +1,203 @@
+    # it creates a simulatioin input object that has dynamic shape If you use it with other gmxapi components.
+    # they understand what it's shape is.   And it does support array indexing. 
+    # once you have createad aan API object. We want people to recognize it's an abstraction. 
+    # should make assumptions about accessing as an array
```
To clarify, we should be clear that the object is not a native Python `list` or `tuple`, or even numpy array, even though we can allow subscripted indexing. This is somewhat clarified at #28 but also scattered around a bit... Unless there are objections, I'll give #28 another pass and then merge it, then try to make a sweep to refresh the various issues to clarify/coordinate discussions on object model, data model, and software components.

eirrgang commented:
> Did you want to give a version that iterated differently?

Yes, sorry. I'll try to do that before next week, but I might not be able to get to it until Friday.

> Fill out more of the operations of the wrapper objects?

I'd be curious to see a LAMMPS version, if you can come up with it.

> Start doing more things with MSM builder?

If you have specific ideas, please share them, but I think the goal here was to show a particular simulation/analysis feedback loop. I think it would be instructive to either show how the same high level operation would be implemented in terms of different tools, or to instead demonstrate other high level operations that are different enough to demonstrate distinct functional requirements.

mrshirts commented:
Addressed issues above and split the code up into independent files (some ambiguity where some functions should go).  lanmmps_wrapper.py mostly a copy of gromacs_wrapper.py at this point.

mrshirts commented:
Questions: 

- LAMMPS inputs are in the form of a single main file, which has commands to include a number of subsidiary files. How do you suggest having such objects be passed in to a routine that does simulation setup?

- Also: most programs don't have an intermediate .tpr file; you launch the simulation directly from the input files (essentially imagine launching mdrun directly from the .gro, .top, .mdp files.  What sort of object should a make_input method return?  Simply a data structure that stores these input files, and can be passed to the simulate method?

eirrgang commented:
> * LAMMPS inputs are in the form of a single main file, which has commands to include a number of subsidiary files. How do you suggest having such objects be passed in to a routine that does simulation setup?

If there are not already Python tools (or even command line tools) to manage this hierarchy of files, then we probably have to require the user to specify all of them, at least initially...

We can develop some amount of tool-specific utilities/adapters for specific packages, but we should decide how much time/effort we want to allocate before forming ideas of the sort of technical solution to pursue; and we should pursue collaboration or co-development of API interoperability with the tool's project team before committing to a lot...

Is there another LAMMPS-wrapping workflow tool that we can draw inspiration from?

> * Also: most programs don't have an intermediate .tpr file; you launch the simulation directly from the input files (essentially imagine launching mdrun directly from the .gro, .top, .mdp files.  What sort of object should a make_input method return? 

The return type of a `make_input` method should be determined by the module associated with the simulator implementation for which it is implemented. (In a future project phase, we can consider a generalization / wrapper / dispatching scheme, but only after we have enough sense of how to interconvert simulation inputs that there would be any utility in the abstraction.)

> Simply a data structure that stores these input files, and can be passed to the simulate method?

That should be sufficient for prototyping.

mrshirts commented:
> If there are not already Python tools (or even command line tools) to manage this hierarchy of files, then we probably have to require the user to specify all of them, at least initially...

I mean, the user specifies these files to be included (roughly, does the same thing as an #include preprocessing directive) in the main input file. So they already specify them, within that file.  That's the tools, a file parser. 

The simplest for the user seems to be could reimplement the parts of the parsing that recognize the files (literally just scanning through the input file for keywords). That would be the simplest for the user.  But it's problematic for handing in a list of files - that would need to be done outside of the input file.

There are no really widely used Python scripting tools.  https://pizza.sandia.gov/ has some Python tools (Pizza.py), but it's not really designed for managing and launching tools. 

> Pursue collaboration or co-development of API interoperability with the tool's project team before committing to a lot...

I'm not aware of any official LAMMPS efforts to develop APIs right now. I think it's more about how can we come up with a consistent way of handling the input/output files they provide at a file i/o level.

> We can develop some amount of tool-specific utilities/adapters for specific packages, but we should decide how much time/effort we want to allocate before forming ideas of the sort of technical solution to pursue;

Well, we need a way to modify input files, or we can't do anything else.

Since LAMMPS uses a basic (more or less) keyword/command syntax, then it presumably would be simple enough to have users pass into modify_inputs a dictionary of keywords that are equal to the lammps keywords, with an array of N possibilities for the options that can be put into N new files in place of the current values, and return an array of new input files.

eirrgang commented:
I'm sorry, but it's been a long time since I've used LAMMPS... I'll have a chance to get back up to speed after next week.

I see that the input file you are talking about is essentially a script for a LAMMPS-specific parser. I don't think it would be worth more than a couple of hours for use to write I/O or parsing for that file. We could do a little bit of basic transformation like you suggest, but the immediate prototyping solution is probably to have the user provide the "script" input file (which we handle opaquely) and declare the important things we need to know about what is in it, like what additional local files we need to consider to be input and what output files to look for.

Then, we should focus our coding against the interfaces LAMMPS already provides. For instance: https://lammps.sandia.gov/doc/Python_head.html

We can offer users the choice of
1. treating LAMMPS as an opaque tool to use in a bag-of-tasks with a user-declared input and output bundle, as with our generic command line wrapper, or
2. using a more uniform SCALE-MS style interface with our full feature set, going as directly as possible from our Python interface to the native tool API, and bypassing the shell and parsed LAMMPS interface.

If it is clear to you that it would be straightforward to compose the LAMMPS input file in terms of some simple building blocks, then that could be a good starting point and should be easy to migrate to a more integrated long term approach. But I think we should avoid trying to convert back and forth (i.e. parsing the LAMMPS file, processing it in Python, then rewriting it). That would be reinventing a wheel and adding a layer where we should be removing one.

Even LAMMPS claims not to be able to parse an input file all at once (it is interpreted line-by-line), so I'd be very cautious about accepting such files as normative SCALE-MS input.

eirrgang commented:
Actually, I have another idea that might be of general utility. We could treat the LAMMPS input file as a sort of generic scalems.TextDocument object. If a raw string is provided with formatting mark-up (e.g. `{}` appearing in Python raw strings), we could apply `str.format` at run time using other scalems objects, which could format as filenames, where applicable. LAMMPS script building blocks could then also be implemented naively as unformatted blocks of lines that can be composed (concatenated) together.

That could provide a simpler migration path for people who have existing LAMMPS scripts. People might even already be composing such files with a variety of macro/substitution methods for filenames.

mrshirts commented:
> We could treat the LAMMPS input file as a sort of generic scalems.TextDocument object. If a raw string is provided with formatting mark-up (e.g. {} appearing in Python raw strings),

That does make a lot of sense.

eirrgang commented:
I updated #24 in response to this PR and additional conversations. I'll make a few comments here about some things that I didn't apply.

It may be appropriate to merge #24 and reduce this PR to a LAMMPS extension of the example in #24 or other additions/corrections that need more granular review.

I will also propose some syntax for scalems.Text interactions in a separate PR.

eirrgang commented:
Is this branch still active? It seems like most of the changes have been merged and that the remaining differences are being pursued in other branches. Can you confirm that this PR should be closed? Or is it useful to keep it open?

eirrgang commented on src/lammps_wrapper.py:
```diff
@@ -0,0 +1,67 @@
+
+import scalems
+
+# NEED TO REWRITE FOR LAMMPS SYNTAX
```
For more discussion, see #37
